### ISSUE TITLE
Different result of std::process::Command's example on different platform

### ISSUE THREAD

**Issue by tiide** [rust-lang/rust#70557]
Summary
---
The example is located at https://doc.rust-lang.org/std/process/struct.Stdio.html.   
When running on my CentOS, **assert** happens **due to an extra character '\n'** at the left side. It exists under both stable and nightly  versions (seeing "My platform info").

Details
---
``` rust
use std::io::Write;
use std::process::{Command, Stdio};

let mut child = Command::new("rev")
    .stdin(Stdio::piped())
    .stdout(Stdio::piped())
    .spawn()
    .expect("Failed to spawn child process");

{
    let stdin = child.stdin.as_mut().expect("Failed to open stdin");
    stdin.write_all("Hello, world!".as_bytes()).expect("Failed to write to stdin");
}

let output = child.wait_with_output().expect("Failed to read stdout");
assert_eq!(String::from_utf8_lossy(&output.stdout), "!dlrow ,olleH");
```

When I click run button, it runs on play.rust-lang.org, nothing assert_eq is triggered as below:   
[run on play][play]
>  Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.65s
     Running `target/debug/playground`

But when I copy the codes to my platform, the assert_eq is triggered as below:
`$ cargo run`
>    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
     Running `target/debug/combine`
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `"!dlrow ,olleH\n"`,
 right: `"!dlrow ,olleH"`', src/main.rs:18:1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

My platform info
==
Name: VMware® Workstation 15 Pro; Ver: 15.5.0 build-14665864
`$ cargo -V`
> cargo 1.42.0 (86334295e 2020-01-31)

`$ rustc -V`
> rustc 1.42.0 (b8cedc004 2020-03-09) 

`$ rustc -V`
> rustc 1.44.0-nightly (f509b26a7 2020-03-18)

`$ cat /proc/version`
> Linux version 3.10.0-1062.12.1.el7.x86_64 (mockbuild@kbuilder.bsys.centos.org) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC) ) #1 SMP Tue Feb 4 23:02:59 UTC 2020

`$ lsb_release -a`
> LSB Version:	:core-4.1-amd64:core-4.1-noarch:cxx-4.1-amd64:cxx-4.1-noarch:desktop-4.1-amd64:desktop-4.1-noarch:languages-4.1-amd64:languages-4.1-noarch:printing-4.1-amd64:printing-4.1-noarch
Distributor ID:	CentOS
Description:	CentOS Linux release 7.7.1908 (Core)
Release:	7.7.1908
Codename:	Core

[play]: https://play.rust-lang.org/?code=%23!%5Ballow(unused)%5D%0Afn%20main()%20%7B%0Ause%20std%3A%3Aio%3A%3AWrite%3B%0Ause%20std%3A%3Aprocess%3A%3A%7BCommand%2C%20Stdio%7D%3B%0A%0Alet%20mut%20child%20%3D%20Command%3A%3Anew(%22rev%22)%0A%20%20%20%20.stdin(Stdio%3A%3Apiped())%0A%20%20%20%20.stdout(Stdio%3A%3Apiped())%0A%20%20%20%20.spawn()%0A%20%20%20%20.expect(%22Failed%20to%20spawn%20child%20process%22)%3B%0A%0A%7B%0A%20%20%20%20let%20stdin%20%3D%20child.stdin.as_mut().expect(%22Failed%20to%20open%20stdin%22)%3B%0A%20%20%20%20stdin.write_all(%22Hello%2C%20world!%22.as_bytes()).expect(%22Failed%20to%20write%20to%20stdin%22)%3B%0A%7D%0A%0Alet%20output%20%3D%20child.wait_with_output().expect(%22Failed%20to%20read%20stdout%22)%3B%0Aassert_eq!(String%3A%3Afrom_utf8_lossy(%26output.stdout)%2C%20%22!dlrow%20%2ColleH%22)%3B%0A%7D&edition=2018

**Comment by jonas-schievink** [comment]
Are you sure this is not a difference in the installed `rev` binaries themselves? (or in the kernel, even?)

**Comment by RalfJung** [comment]
On my system, `echo -n 'Hello, world!' | rev` has no trailing newline. Notice how it ends up on the same line as my shell prompt:
```
$ echo -n 'Hello, world!' | rev
!dlrow ,olleHr@mycomputername$ 
```

**Comment by Mark-Simulacrum** [comment]
Notably, on macOS, I *am* getting a newline:

```
$ echo -n 'Hello, world!' | rev | xxd # on macOS
00000000: 2164 6c72 6f77 202c 6f6c 6c65 480a       !dlrow ,olleH.

$ echo -n 'Hello, world!' | rev | xxd # on Linux
00000000: 2164 6c72 6f77 202c 6f6c 6c65 48         !dlrow ,olleH
```

I'm pretty sure this is not a bug in Rust.

**Comment by tiide** [comment]
```bash
$  echo -n 'Hello, world!' | rev | xxd 
00000000: 2164 6c72 6f77 202c 6f6c 6c65 480a       !dlrow ,olleH.
```
I try the shell command, and from the result with a trailing newline, also pretty sure it is really up to the platform or rev itself, nothing to do with Rust.

That's good, thanks.
